fix unique schema key issues

zedrdave opened the pull request:
Fixed 2 issues with current code:

```py
        try:
            hash(attribute)
        except TypeError:
            attribute = tuple(attribute)
```

1. If `attribute` is immutable (and can be hashed) nothing is actually done (the hash is not assigned).

2. For schema attributes that are `set` (such as `only`, `exclude` etc), an exception is triggered and a tuple is used, which randomly fails to match on itself (since `str(tuple(x))` is non-deterministic when `x` is a `set`)…

lafrech committed:
Fix test_schema_array_uses_ref_if_available

lafrech committed:
Fix reference parameter/response with MarshmallowPlugin

Commit:
fix unique schema key issues

Commit:
fix unique schema key issues

dependabot-support committed:
Bump pre-commit from 1.14.4 to 1.15.0

Bumps [pre-commit](https://github.com/pre-commit/pre-commit) from 1.14.4 to 1.15.0.
- [Release notes](https://github.com/pre-commit/pre-commit/releases)
- [Changelog](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pre-commit/pre-commit/compare/v1.14.4...v1.15.0)

Signed-off-by: dependabot[bot] <support@dependabot.com>

lafrech commented:
Thanks for submitting this. I think it is on the right track.

We could discuss what needs to be covered by a `try`/`except` and by a `if` condition, but on the principle, it looks good to me.

Do you think you could add tests?

dependabot-support committed:
Bump pre-commit from 1.15.0 to 1.15.1

Bumps [pre-commit](https://github.com/pre-commit/pre-commit) from 1.15.0 to 1.15.1.
- [Release notes](https://github.com/pre-commit/pre-commit/releases)
- [Changelog](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pre-commit/pre-commit/compare/v1.15.0...v1.15.1)

Signed-off-by: dependabot[bot] <support@dependabot.com>

lafrech commented:
```py
MODIFIERS = ["only", "exclude", "load_only", "dump_only", "partial"]
```

Modifiers can be strings or iterables (hashable or not) such as list, tuple, set.

Would the code be simpler and the intent clearer if we just wrote it like this?

```py
        try:
            # Hashable (string, tuple)
            attribute = hash(attribute)
        except TypeError:
            # Unhashable iterable (list, set)
            attribute = hash(frozenset(attribute))
```

lafrech committed:
Factorize COMPONENT_SUBSECTIONS and reference_path in utils.py

lafrech committed:
Change utils.reference_path into utils.build_reference

lafrech committed:
Change tests.utils.ref_path into tests.utils.build_ref

lafrech committed:
Merge pull request #414 from marshmallow-code/fix_resolve_parameters_resolve_schema

 Fix reference parameter/response with MarshmallowPlugin

lafrech committed:
Merge pull request #419 from marshmallow-code/factorize_component_subsections_mapping

Factorize components subsections mapping and add utils.build_ref

lafrech committed:
Strip empty component sections from doc

lafrech committed:
Update CHANGELOG

sloria committed:
Bump version and update changelog

lafrech committed:
Fix release date typo in CHANGELOG.rst

Bangertm commented:
Is there an example of a case where classes with the same modifiers are not matching properly?

The original intent of `make_schema_key` was to create a key that can be used in the `ref` dictionary.  We were originally using just the schema class for the key but when we wanted to take modifiers into account we need a way to include them in that key.  My thought process was to create a NamedTuple that included the class and each of the modifiers, but since some of the modifiers are not hashable (sets) we needed to convert them to something that is hashable.

The intent behind this code block is to test to see if the attribute is hashable and if not make it hashable.  I wanted to avoid storing the hash in the NamedTuple for debug reasons i.e. if someone wanted to inspect the `ref` dictionary it would be more convenient to see something that looked like the original modifier than a random number.  

Maybe instead of changing unhasable attributes to tuples we just change them to frozensets?

lafrech commented:
> Maybe instead of changing unhasable attributes to tuples we just change them to frozensets?

Good point.

```python
        try:
            # Hashable (string, tuple)
            hash(attribute)
        except TypeError:
            # Unhashable iterable (list, set)
            attribute = frozenset(attribute)
```

lafrech committed:
Coerce response status codes to strings

sloria committed:
Merge pull request #424 from marshmallow-code/coerce_status_code_string

Coerce response status codes to strings

sloria committed:
Bump version and update changelog

lafrech committed:
tests.utils.get_*: no default value on missing item

lafrech committed:
Merge pull request #421 from marshmallow-code/strip_empty_subsections_from_docs

Strip empty component sections from doc

lafrech committed:
Strip tags from doc if empty

lafrech committed:
Merge pull request #425 from marshmallow-code/strip_empty_tags_from_docs

Strip tags from doc if empty

lafrech committed:
Bump version and update changelog

dependabot-support committed:
Bump sphinx from 2.0.0 to 2.0.1

Bumps [sphinx](https://github.com/sphinx-doc/sphinx) from 2.0.0 to 2.0.1.
- [Release notes](https://github.com/sphinx-doc/sphinx/releases)
- [Changelog](https://github.com/sphinx-doc/sphinx/blob/master/CHANGES)
- [Commits](https://github.com/sphinx-doc/sphinx/compare/v2.0.0...v2.0.1)

Signed-off-by: dependabot[bot] <support@dependabot.com>

lafrech commented:
I think my last proposal is in line with @Bangertm's intent and fixes the bug.

@zedrdave, would you like to

- update your PR 
- add a test case
- add yourself to AUTHORS.rst

?

Commit:
Fix Marshmallow Schema's Meta exclude field list/tuple error

Commit:
Use pytest.mark consistently

There are many uses of pytest.mark and two of them use the imported mark symbol. Most
uses access it via pytest.mark. This changes the two inconsistent references to also
use pytest.mark.

Commit:
Parameterize the validation property tests

Rather than test each field validation property with a separate test, each
field can have a set of properties that gets checked.

Commit:
Reduce scope of ValidationSchema to the test class

dependabot-support committed:
Bump pre-commit from 1.15.1 to 1.15.2

Bumps [pre-commit](https://github.com/pre-commit/pre-commit) from 1.15.1 to 1.15.2.
- [Release notes](https://github.com/pre-commit/pre-commit/releases)
- [Changelog](https://github.com/pre-commit/pre-commit/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pre-commit/pre-commit/compare/v1.15.1...v1.15.2)

Signed-off-by: dependabot[bot] <support@dependabot.com>

Commit:
Find type and format for field using type hierarchy

When a custom field is not directly found in the field mapping then we use
the MRO to look at parent classes to find  a suitable class to work out
what type of field this is.

Commit:
Add Andrew Johnson to authors list

lafrech committed:
Rework ext.marshmallow.common.get_fields tests

lafrech committed:
Merge pull request #431 from blagasz/dev

Fix Marshmallow Schema's Meta exclude field error

lafrech committed:
Bump version and update changelog

sloria committed:
Merge pull request #435 from andrjohn/lists-are-arrays

Uses class hierarchy to find type and format properties

sloria committed:
Fix watch-docs

sloria committed:
Bump version and update changelog

lafrech committed:
tox/pre-commit: show-diff-on-failure

sloria committed:
Merge pull request #438 from marshmallow-code/pre-commit-show-diff-on-failure

tox/pre-commit: show-diff-on-failure

zedrdave commented:
@lafrech Hi, I finally have a bit of spare time. Will try and take care of this today.

Commit:
Merge branch 'fix-make-schema-key' of github.com:zedrdave/apispec into fix-make-schema-key

Commit:
schema key with test

lafrech commented on tests/test_ext_marshmallow_common.py:
```diff
@@ -19,13 +19,19 @@ def test_raise_if_schema_class_passed(self):
+        # Note: this test relies on non-deterministic hashing behaviour:
+        for i in range(0, 50):
+            assert make_schema_key(PetSchema(load_only=('1', '2', '3', '4'))) == make_schema_key(PetSchema(load_only=('4', '3', '2', '1')))
+
```
Those are not sets but tuples, so I don't think this test tests what it means to.

I understand the reason for the loop, but I wouldn't do that. I'd just test with a set and a loop.

```py
assert make_schema_key(PetSchema(load_only=['1', '2', '3', '4'])) == make_schema_key(PetSchema(load_only=['4', '3', '2', '1']))
assert make_schema_key(PetSchema(load_only=set(['1', '2', '3', '4']))) == make_schema_key(PetSchema(load_only=set(['4', '3', '2', '1'])))
```

lafrech commented:
Do you think you could rebase on dev (rather than merge dev in your branch)? It would make the diff and the history more readable. At this point, you're better off starting a new branch from dev and cherry-picking your fix. You can push force in this branch or start a new PR if that's more convenient to you.

I that's too much trouble, I'll try to find the time to pick this up.

lafrech commented:
I did the rework in https://github.com/marshmallow-code/apispec/pull/439.

zedrdave commented:
@lafrech Thanks for doing the rewrite! (sorry for the lack of reactivity… haven't had lots of time to work on side-projects lately)
